txscript/script.go

```diff
@@ -420,6 +446,216 @@ func CalcWitnessSigHash(script []byte, sigHashes *TxSigHashes, hType SigHashType
+	// to the sighash digest message. This is used for segwit v1 spends,
+	// a.k.a the tapscript keyspend path.
+	baseSigHashExtFlag sigHashExtFlag = 0
+)
+
+// taprootSigHashOptions houses a set of functional options that may optionally
+// modify how the taproot/script sighash digest algorithm is implemented.
+type taprootSigHashOptions struct {
+}
+// defaultTaprootSighashOptions returns the set of default sighash options for
+// taproot execution.
+func defaultTaprootSighashOptions() *taprootSigHashOptions {
+	return &taprootSigHashOptions{}
+}
+
+// TaprootSigHashOption defines a set of functional param options that can be
+// used to modify the base sighash message with optional extensions.
+type TaprootSigHashOption func(*taprootSigHashOptions)
+// calcTaprootSignatureHashRaw computes the sighash as specified in BIP 143.
+// If an invalid sighash type is passed in, an error is returned.
+func calcTaprootSignatureHashRaw(sigHashes *TxSigHashes, hType SigHashType,
+	tx *wire.MsgTx, idx int,
+	prevOutFetcher PrevOutputFetcher,
+	sigHashOpts ...TaprootSigHashOption) ([]byte, error) {
+
+	opts := defaultTaprootSighashOptions()
+	for _, sigHashOpt := range sigHashOpts {
+		sigHashOpt(opts)
+	}
+
+	// If a valid sighash type isn't passed in, then we'll exit early.
+	if !isValidTaprootSigHash(hType) {
+		// TODO(roasbeef): use actual errr here
+		return nil, fmt.Errorf("invalid taproot sighash type: %v", hType)
+	}
+
+	// As a sanity check, ensure the passed input index for the transaction
+	// is valid.
+	if idx > len(tx.TxIn)-1 {
+		return nil, fmt.Errorf("idx %d but %d txins", idx, len(tx.TxIn))
+	}
+
+	// We'll utilize this buffer throughout to incrementally calculate
+	// the signature hash for this transaction.
+	var sigMsg bytes.Buffer
+
+	// The final sighash always has a value of 0x00 prepended to it, which
+	// is called the sighash epoch.
+	sigMsg.WriteByte(0x00)
+
+	// First, we write the hash type encoded as a single byte.
+	if err := sigMsg.WriteByte(byte(hType)); err != nil {
+		return nil, err
+	}
+
+	// Next we'll write out the transaction specific data which binds the
+	// outer context of the sighash.
+	err := binary.Write(&sigMsg, binary.LittleEndian, tx.Version)
+	if err != nil {
+		return nil, err
+	}
+	err = binary.Write(&sigMsg, binary.LittleEndian, tx.LockTime)
+	if err != nil {
+		return nil, err
+	}
+
+	// If sighash isn't anyone can pay, then we'll include all the
+	// pre-computed midstate digests in the sighash.
+	if hType&SigHashAnyOneCanPay != SigHashAnyOneCanPay {
+		sigMsg.Write(sigHashes.HashPrevOutsV1[:])
+		sigMsg.Write(sigHashes.HashInputAmountsV1[:])
+		sigMsg.Write(sigHashes.HashInputScriptsV1[:])
+		sigMsg.Write(sigHashes.HashSequenceV1[:])
+	}
+
+	// If this is sighash all, or its taproot alias (sig hash default),
+	// then we'll also include the pre-computed digest of all the outputs
+	// of the transaction.
+	if hType == SigHashDefault || hType == SigHashAll {
```

Comment:
IMO the BIP text here is confusing as it states the sighash types by process of elimination rather than being more explicit, but updated to match more directly. It directs implementation to use a bitmask when it isn't actually necessary as you can just compare the raw types, unlike sighash anyone. 

Nevertheless, changed to match the BIP.